read-fonts/src/tables/truetype/mem.rs

```diff
@@ -0,0 +1,188 @@
+//! Memory allocation for TrueType scaling.
+
+use std::mem::{align_of, size_of};
+
+use types::{F26Dot6, Fixed, Point};
+
+use super::{super::glyf::PointFlags, OutlineInfo};
+
+/// Buffers used during outline scaling.
+pub struct OutlineMemory<'a> {
+    pub unscaled: &'a mut [Point<i32>],
+    pub scaled: &'a mut [Point<F26Dot6>],
+    pub original_scaled: &'a mut [Point<F26Dot6>],
+    pub contours: &'a mut [u16],
+    pub flags: &'a mut [PointFlags],
+    pub deltas: &'a mut [Point<Fixed>],
+    pub iup_buffer: &'a mut [Point<Fixed>],
+    pub composite_deltas: &'a mut [Point<Fixed>],
+}
+
+impl<'a> OutlineMemory<'a> {
+    pub fn new(info: &OutlineInfo, buf: &'a mut [u8]) -> Option<Self> {
+        let (scaled, buf) = alloc_slice(buf, info.points)?;
+        let (unscaled, buf) = alloc_slice(buf, info.max_other_points)?;
+        // We only need original scaled points when hinting
+        let (original_scaled, buf) = if info.has_hinting {
+            alloc_slice(buf, info.max_other_points)?
+        } else {
+            (Default::default(), buf)
+        };
+        // Don't allocate any delta buffers if we don't have variations
+        let (deltas, iup_buffer, composite_deltas, buf) = if info.has_variations {
+            let (deltas, buf) = alloc_slice(buf, info.max_simple_points)?;
+            let (iup_buffer, buf) = alloc_slice(buf, info.max_simple_points)?;
+            let (composite_deltas, buf) = alloc_slice(buf, info.max_component_delta_stack)?;
+            (deltas, iup_buffer, composite_deltas, buf)
+        } else {
+            (
+                Default::default(),
+                Default::default(),
+                Default::default(),
+                buf,
+            )
+        };
+        let (contours, buf) = alloc_slice(buf, info.contours)?;
+        let (flags, _) = alloc_slice(buf, info.points)?;
+        Some(Self {
+            unscaled,
+            scaled,
+            original_scaled,
+            contours,
+            flags,
+            deltas,
+            iup_buffer,
+            composite_deltas,
+        })
+    }
+}
+
+/// Trait that defines which types may be used as element types for
+/// `alloc_slice`.
+///
+/// # Safety
+/// This must only be implemented for types that contain no internal padding
+/// and for which all bit patterns are valid.
+unsafe trait TransmuteElement: Sized + Copy {}
+
+unsafe impl TransmuteElement for u16 {}
+unsafe impl TransmuteElement for i32 {}
+unsafe impl TransmuteElement for Fixed {}
+unsafe impl TransmuteElement for F26Dot6 {}
+unsafe impl TransmuteElement for PointFlags {}
```

Comment:
I'd need a second opinion on this, but I'm not sure if `PointFlags` is safe. I would mark it `#[repr(transparent)]` to guarantee that it has the same layout as a `u8` (although obviously it should??)

Reply:
Another good catch. Will add `repr(transparent)`

Reply:
Done. Also added this to the fixed point types. I thought we had already done it for some reason.